### PR TITLE
derive Serialize and Deserialize for XCM primitives

### LINKED ISSUE #4930: derive Serialize and Deserialize for XCM primitives
It could be useful for XCM primitives, such as ```MultiLocation``` and ```MultiAsset```, to have ```Serialize``` and ```Deserialize``` impl. This will allow them to be stored in and to be loaded from genesis file.

### PR DESCRIPTION
**Pull request by stanly-johnson**
resolves #4930

### CODE REVIEW

**Comment by bkchr** [comment]
@stanly-johnson did you even test this? As you see the entire CI is failing.

File: xcm/src/v1/multiasset.rs [hunk 1]
```diff
@@ -32,9 +32,12 @@ use core::{
 };
 use parity_scale_codec::{self as codec, Decode, Encode};
 use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
```
Comment (bkchr) [inline xcm/src/v1/multiasset.rs:36]:
```suggestion
```

File: xcm/src/v1/multiasset.rs [hunk 2]
```diff
@@ -32,9 +32,12 @@ use core::{
...
 
 /// A general identifier for an instance of a non-fungible asset class.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v1/multiasset.rs:40]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v1/multiasset.rs [hunk 3]
```diff
@@ -97,6 +100,7 @@ impl From<Vec<u8>> for AssetInstance {
 
 /// Classification of an asset being concrete or abstract.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Encode, Decode, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v1/multiasset.rs:103]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v1/multiasset.rs [hunk 4]
```diff
@@ -146,6 +150,7 @@ impl AssetId {
 
 /// Classification of whether an asset is fungible or not, along with a mandatory amount or instance.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Debug, Encode, Decode, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v1/multiasset.rs:153]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v1/multiasset.rs [hunk 5]
```diff
@@ -173,6 +178,7 @@ impl<T: Into<AssetInstance>> From<T> for Fungibility {
 }
 
 #[derive(Clone, Eq, PartialEq, Debug, Encode, Decode, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v1/multiasset.rs:181]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v1/multilocation.rs [hunk 6]
```diff
@@ -48,6 +50,7 @@ use scale_info::TypeInfo;
 ///
 /// The `MultiLocation` value of `Null` simply refers to the interpreting consensus system.
 #[derive(Clone, Decode, Encode, Eq, PartialEq, Ord, PartialOrd, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v1/multilocation.rs:53]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v1/multilocation.rs [hunk 7]
```diff
@@ -426,6 +429,7 @@ const MAX_JUNCTIONS: usize = 8;
 /// Parent junctions cannot be constructed with this type. Refer to `MultiLocation` for
 /// instructions on constructing parent junctions.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v1/multilocation.rs:432]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v1/multilocation.rs [hunk 8]
```diff
@@ -20,6 +20,8 @@ use super::Junction;
 use core::{convert::TryFrom, mem, result};
 use parity_scale_codec::{Decode, Encode};
 use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
```
Comment (bkchr) [inline xcm/src/v1/multilocation.rs:24]:
```suggestion
```

File: xcm/src/v1/junction.rs [hunk 9]
```diff
@@ -22,11 +22,14 @@ use alloc::vec::Vec;
 use core::convert::TryFrom;
 use parity_scale_codec::{self, Decode, Encode};
 use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
```
Comment (bkchr) [inline xcm/src/v1/junction.rs:26]:
```suggestion
```

File: xcm/src/v1/junction.rs [hunk 10]
```diff
@@ -22,11 +22,14 @@ use alloc::vec::Vec;
...
 
 /// A single item in a path to describe the relative location of a consensus system.
 ///
 /// Each item assumes a pre-existing location as its context and is defined in terms of it.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v1/junction.rs:32]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v0/multi_location.rs [hunk 11]
```diff
@@ -19,6 +19,8 @@
 use super::Junction;
 use core::{mem, result};
 use parity_scale_codec::{self, Decode, Encode};
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
```
Comment (bkchr) [inline xcm/src/v0/multi_location.rs:23]:
```suggestion
```

File: xcm/src/v0/multi_location.rs [hunk 12]
```diff
@@ -43,6 +45,7 @@ use parity_scale_codec::{self, Decode, Encode};
 ///
 /// The `MultiLocation` value of `Null` simply refers to the interpreting consensus system.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, scale_info::TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v0/multi_location.rs:48]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v0/multi_asset.rs [hunk 13]
```diff
@@ -25,6 +25,8 @@ use core::{
 };
 use parity_scale_codec::{self, Decode, Encode};
 use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
```
Comment (bkchr) [inline xcm/src/v0/multi_asset.rs:29]:
```suggestion
```

File: xcm/src/v0/multi_asset.rs [hunk 14]
```diff
@@ -84,6 +86,7 @@ pub use crate::v1::AssetInstance;
 /// - `<chain>/AccountKey20` for an ERC-20-style single-asset smart-contract on an Ethereum-like chain.
 ///
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v0/multi_asset.rs:89]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v0/junction.rs [hunk 15]
```diff
@@ -19,9 +19,12 @@
 use alloc::vec::Vec;
 use parity_scale_codec::{Decode, Encode};
 use scale_info::TypeInfo;
+#[cfg(feature = "std")]
+use serde::{Deserialize, Serialize};
```
Comment (bkchr) [inline xcm/src/v0/junction.rs:23]:
```suggestion
```

File: xcm/src/v0/junction.rs [hunk 16]
```diff
@@ -19,9 +19,12 @@
...
 
 /// A global identifier of an account-bearing consensus system.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v0/junction.rs:27]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v0/junction.rs [hunk 17]
```diff
@@ -35,6 +38,7 @@ pub enum NetworkId {
 
 /// An identifier of a pluralistic body.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v0/junction.rs:41]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v0/junction.rs [hunk 18]
```diff
@@ -56,6 +60,7 @@ pub enum BodyId {
 
 /// A part of a pluralistic body.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v0/junction.rs:63]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

File: xcm/src/v0/junction.rs [hunk 19]
```diff
@@ -103,6 +108,7 @@ impl BodyPart {
 ///
 /// Each item assumes a pre-existing location as its context and is defined in terms of it.
 #[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug, TypeInfo)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
```
Comment (bkchr) [inline xcm/src/v0/junction.rs:111]:
```suggestion
#[cfg_attr(feature = "std", derive(serde::Serialize, serde::Deserialize))]
```

**Review by bkchr — COMMENTED** [review]
It should probably also directly target the XCMV3 branch. WDYT @KiChjang ?

**Comment by KiChjang** [comment]
> It should probably also directly target the XCMV3 branch. WDYT @KiChjang ?

Yes, this should happen as well, so that Gav may have a chance to look at it and be aware of the changes made here.

**Comment by stanly-johnson** [comment]
messed up the rebase, replaced by #5036

**Pull request closed without merging**